### PR TITLE
WIP: Add ant ivy scripts 2

### LINKED ISSUE #45: add ant skripts
add ant build skripts w/ same functionality as the already existing gradle scripts and the future maven skripts (#38, #39)

### LINKED ISSUE #61: Doc for Ant
for #57

### PR DESCRIPTION
**Pull request by Tobias321**
fixes #45, fixes #61 

this is still WIP

### CODE REVIEW

File: build.properties [hunk 1]
```diff
@@ -0,0 +1,3 @@
+gsonVersion=2.8.6
+gdxVersion=1.9.8
```
Comment (cagix) [inline build.properties:2]: in maven (#58) this is `1.9.12`?

File: build.xml [hunk 2]
```diff
@@ -0,0 +1,84 @@
+<project xmlns:ivy="antlib:org.apache.ivy.ant" name="pmdungeon" basedir="." default="clean-build">
+    <property name="build.dir" value="build" />
+    <property name="lib.dir" value="libs" />
+    <property name="classes.dir" value="${build.dir}/classes" />
+    <property name="jar.dir" value="${build.dir}/jar" />
+
+    <property name="ivy.install.version" value="2.1.0-rc2" />
```
Comment (cagix) [inline build.xml:7]: could the ivy version be defined in `build.properties` as well?

**Review by cagix — CHANGES_REQUESTED** [review]
looks promising 👍

**Comment by Tobias321** [comment]
i have added a ivyVersion property.

i don't understand, why now 8 files have changed. that wasn't an intention.

**Comment by Tobias321** [comment]
@cagix side note: you have to take care, i started the rebase via cmd... but solved the conflicts via github desktop. then i wanted to push via gd too, but github desktop then performs a "pull-push" (to avoid fast-forwarding, i assume...)

**Comment by cagix** [comment]
> @cagix side note: you have to take care, i started the rebase via cmd... but solved the conflicts via github desktop. then i wanted to push via gd too, but github desktop then performs a "pull-push" (to avoid fast-forwarding, i assume...)

i really start to dislike this tool ...

**Comment by Tobias321** [comment]
have a look at `https://repo1.maven.org/maven2/com/badlogicgames/gdx/gdx-platform/1.9.12/`
which of the listed dependencies could we need?
i tried it with "all":
```xml
    <dependencies>
        <dependency org="com.google.code.gson" name="gson" rev="${gsonVersion}" />
        <dependency org="com.badlogicgames.gdx" name="gdx-ai" rev="${gdxAiVersion}" />
        <dependency org="com.badlogicgames.gdx" name="gdx" rev="${gdxVersion}" />
        <dependency org="com.badlogicgames.gdx" name="gdx-backend-lwjgl" rev="${gdxVersion}" />
        <dependency org="com.badlogicgames.gdx" name="gdx-backend-lwjgl3" rev="${gdxVersion}" />
        <dependency org="com.badlogicgames.gdx" name="gdx-platform" rev="${gdxVersion}">
            <artifact name="gdx-platform" type="jar" />
            <artifact name="gdx-platform" type="jar" e:classifier="natives-desktop" />
            <artifact name="gdx-platform" type="jar" e:classifier="natives-x86" />
            <artifact name="gdx-platform" type="jar" e:classifier="natives-x86_64" />
            <artifact name="gdx-platform" type="jar" e:classifier="natives-arm64-v8a" />
            <artifact name="gdx-platform" type="jar" e:classifier="natives-ios" />
        </dependency>
        <dependency org="com.badlogicgames.gdx" name="gdx-box2d" rev="${gdxVersion}" />
        <dependency org="com.badlogicgames.gdx" name="gdx-box2d-platform" rev="${gdxVersion}">
            <artifact name="gdx-box2d-platform" type="jar" />
            <artifact name="gdx-box2d-platform" type="jar" e:classifier="natives-desktop" />
            <artifact name="gdx-box2d-platform" type="jar" e:classifier="natives-x86" />
            <artifact name="gdx-box2d-platform" type="jar" e:classifier="natives-x86_64" />
            <artifact name="gdx-box2d-platform" type="jar" e:classifier="natives-arm64-v8a" />
            <artifact name="gdx-box2d-platform" type="jar" e:classifier="natives-ios" />
        </dependency>
        <dependency org="com.badlogicgames.gdx" name="gdx-freetype" rev="${gdxVersion}" />
        <dependency org="com.badlogicgames.gdx" name="gdx-freetype-platform" rev="${gdxVersion}">
            <artifact name="gdx-freetype-platform" type="jar" />
            <artifact name="gdx-freetype-platform" type="jar" e:classifier="natives-desktop" />
            <artifact name="gdx-freetype-platform" type="jar" e:classifier="natives-x86" />
            <artifact name="gdx-freetype-platform" type="jar" e:classifier="natives-x86_64" />
            <artifact name="gdx-freetype-platform" type="jar" e:classifier="natives-arm64-v8a" />
            <artifact name="gdx-freetype-platform" type="jar" e:classifier="natives-ios" />
        </dependency>
        <dependency org="com.badlogicgames.gdx" name="gdx-tools" rev="${gdxVersion}" />
        <!-- <conflict manager="latest-compatible" /> -->
        <conflict manager="all" />
    </dependencies>
```
but there is still the same error, like there's still missing something, that ivy not but maven already retrieve.

**Comment by Tobias321** [comment]
works now 🙂 

i have compared the maven and ivy dependencies. we have to add lwjgl dependency, the natives for gdx-platform and to remove the gdx-tools dependency.

tested on windows x64. please test it too.

**Comment by AMatutat** [comment]
still got
````
compile-run:
     [java] java.lang.NoClassDefFoundError: org/lwjgl/glfw/GLFWErrorCallbackI
```

**Comment by AMatutat** [comment]
Now i get
```
compile-run:
     [java] [LWJGL] Failed to load a library. Possible solutions:
     [java]     a) Add the directory that contains the shared library to -Djava.library.path or -Dorg.lwjgl.librarypath.
     [java]     b) Add the JAR that contains the shared library to the classpath.

```

**Comment by Tobias321** [comment]
@AMatutat can you show me the files of your `pmdungeon/lib` folder? are all natives, lwjgl and opengl jars in there?

or try to run with `-Djava.library.path="..."`

**Comment by AMatutat** [comment]
image

**Comment by Tobias321** [comment]
@AMatutat seems, natives-windows are missing at yours...

image

**Comment by AMatutat** [comment]
@Tobias321 but why?

**Comment by Tobias321** [comment]
> @Tobias321 but why?

i have deleted the `.ant`, `build`, `lib` and `libs` folders. then i have run `ant init-ivy` and `ant compile-run`

**Comment by Tobias321** [comment]
https://me.me/i/wait-see-what-i-found-in-this-room-free-him-c20e8e2b15c9402a83ae392624e4342b 😉

**Comment by AMatutat** [comment]
> i have deleted the .ant, build, lib and libs folders. then i have run ant init-ivy and ant compile-run

works now. Guess my .ant is to blame.

File: README.md [hunk 3]
```diff
@@ -25,7 +25,7 @@ lang: en-EN
 TODO
 
 ### ANT
-- TODO
+- Run `ant init-ivy` once in your project folder, to initialize ivy, then use `ant compile-run` to compile your sources and to start the game.
```
Comment (AMatutat) [inline README.md:28]: something like "You have to add your main class in file blabala line x" is missing

File: build.xml [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+              We will not fail as long as local lib dir exists (it may be empty) and
+              ivy is in at least one of ant's lib dir or the local lib dir. -->
+        <path id="ivy.lib.path">
+            <fileset dir="${ivy.jar.dir}" includes="*.jar" />
+        </path>
+        <taskdef resource="org/apache/ivy/ant/antlib.xml" uri="antlib:org.apache.ivy.ant" classpathref="ivy.lib.path" />
+    </target>
+
+    <target name="resolve" description="target to retrieve all dependencies with Ivy">
+        <!-- <ivy:retrieve pattern="${lib.dir}/[artifact]-[revision](-[classifier]).[ext]" /> -->
```
Comment (AMatutat) [inline build.xml:84]: Can we delete that? I see no reason for commented out code.

File: ivy.xml [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+        <dependency org="com.badlogicgames.gdx" name="gdx-box2d" rev="${gdxVersion}" />
+        <dependency org="com.badlogicgames.gdx" name="gdx-box2d-platform" rev="${gdxVersion}" />
+        <dependency org="com.badlogicgames.gdx" name="gdx-freetype" rev="${gdxVersion}" />
+        <dependency org="com.badlogicgames.gdx" name="gdx-freetype-platform" rev="${gdxVersion}" />
+        <dependency org="org.lwjgl" name="lwjgl" rev="${lwjglVersion}" />
+        <dependency org="org.lwjgl" name="lwjgl-glfw" rev="${lwjglVersion}" />
+        <dependency org="org.lwjgl" name="lwjgl-opengl" rev="${lwjglVersion}" />
+        <dependency org="org.lwjgl" name="lwjgl-jemalloc" rev="${lwjglVersion}" />
+        <dependency org="org.lwjgl" name="lwjgl-openal" rev="${lwjglVersion}" />
+        <!-- <conflict manager="latest-compatible" /> -->
```
Comment (AMatutat) [inline ivy.xml:21]: Can we delete that? I see no reason for commented out code.

File: ivysettings.xml [hunk 6]
```diff
@@ -0,0 +1,10 @@
+<ivysettings>
+    <properties file="build.properties" />
+    <settings defaultResolver="repos" />
+    <resolvers>
+        <chain name="repos">
+            <ibiblio name="my-maven" m2compatible="true" root="https://repo1.maven.org/maven2/" />
+            <!-- <ibiblio name="my-jcenter" m2compatible="true" root="https://jcenter.bintray.com/" /> -->
```
Comment (AMatutat) [inline ivysettings.xml:7]: Can we delete that? I see no reason for commented out code.

**Review by AMatutat — CHANGES_REQUESTED** [review]

File: README.md (hunk 3 above)
Comment (AMatutat) [reply README.md:28]: oh and can we add a link to the Ant page?

**Comment by Tobias321** [comment]
> > i have deleted the .ant, build, lib and libs folders. then i have run ant init-ivy and ant compile-run
> 
> works now. Guess my .ant is to blame.

.ant is the cache folder for ant and ivy's home. sometimes is a good idea, to clear the (entire) cache.

File: README.md (hunk 3 above)
Comment (Tobias321) [reply README.md]: done. but which link do you mean?

**Review by AMatutat — APPROVED** [review]

**Pull request merged**